### PR TITLE
[tests] JavaScript: extract searchindex.js-format test fixtures.

### LINKED ISSUE #12099: Tests: generate the indexes in the HTML search JavaScript tests using the application code.
**Is your feature request related to a problem? Please describe.**
We have JavaScript [test coverage of our HTML search functionality](https://github.com/sphinx-doc/sphinx/blob/5695b76f05af253f8eb3a956d25950af81befc7e/tests/js/searchtools.js), and that's great.

However: the tests use [inline-declared index datastructures](https://github.com/sphinx-doc/sphinx/blob/5695b76f05af253f8eb3a956d25950af81befc7e/tests/js/searchtools.js#L6-L12) to represent what each test index is expected to contain.

That means that the format of the search indexes produced by the application code could go out of sync with the JavaScript tests unnoticed, and we may not catch regressions -- or, equally as annoyingly, it might be very difficult to accurately write additional tests from scratch in future.

**Describe the solution you'd like**
~~Although it would add Python, and `sphinx`, and its requirements to the dependencies for the [JavaScript test workflow](https://github.com/sphinx-doc/sphinx/blob/5695b76f05af253f8eb3a956d25950af81befc7e/.github/workflows/nodejs.yml#L29-L43), I think we should do that and build test index fixtures dynamically.~~

~~They should be checked-in to source control, and the test would fail if the generated-index didn't match the one currently checked-in.~~

An approach that does not require a Python runtime dependency for the JavaScript tests was found instead; the indices can be built using Python, then committed to source control and read as fixture data when the JavaScript tests run.

To catch changes to the index format, a unit test in the Python test suite attempts to regenerate them using the latest application code, and fails when differences are detected in the content.

**Describe alternatives you've considered**
Allowing the status-quo to continue seems fine, but the fact that we lack continuous committers / experts in the search area (it has taken me a while to begin to understand it, and I still don't feel completely confident about it) makes me want to guard it a bit against lack of continuous maintenance/knowledge.

**Additional context**
This would require good documentation too.

- #12028 was a false-alarm where I _thought_ that something about the test index data was wrong -- but it wasn't.  That would have been easier to prove/disprove if the indexes were generated from a project directory.

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Refactoring

### Purpose
- Reformats the test `index` data in the JavaScript HTML search tests to match the format emitted by the Sphinx (Python) HTML builder.

### Detail
- The content of `searchindex.js` is a JavaScript code file that runs `Search.setIndex(...)` with the contents of the index as an argument.
- Instead of loading the index per-test from a variable, retrieve each fixture from the built-in local `karma` HTTP server and `eval` it.
- Includes fixtures generated from genuine, buildable test Sphinx project sources included within this repository.
- Fails the Python test suite when regenerated `searchindex.js` fixture content does not match the existing fixture files.

### Relates
- Resolves #12099.

### CODE REVIEW

**Comment by jayaddison** [comment]
Another challenge: the [`C++` term in an existing test index](https://github.com/sphinx-doc/sphinx/blob/6af611ed85b3de3c5aa5a98bcb8cf9bb5d437aa9/tests/js/searchtools.js#L9) needs to be replicated if we follow this approach.  The current EN-language stemmer maps the input term `C++` to `c` (lowercase).

I don't want to merge this PR until determining a path forward to resolve that, because there might be a conflict otherwise.

Using a null term stemmer could be a solution, but I'd have to figure out how to do that.

**Comment by jayaddison** [comment]
> In follow-up pull requests, the plan is to generate fixtures from genuine, buildable test Sphinx project sources included within this repository.

This turned out to be less code than I expected, so I've bundled it in here too, but could separate it again if that's easier for review/understanding.

File: tests/js/searchtools.js
Comment (jayaddison) [inline tests/js/searchtools.js]: Ah, hm.  I have a suggestion for myself here: load these using the filename `searchindex.js`, but from a directory that has the same name as the source directory under `js/roots`.

That'll be clearer and easier when mapping the source artifacts to their destinations.

**Comment by jayaddison** [comment]
@picnixz when you have time, could you review this pull request next, and then #12107 after that?

I know there are a lot of these small changes / PRs, so it's probably tricky to develop the context for what's changing.  Basically: with these two I want to make it much easier and better to write accurate test coverage, by using 'genuine' generated Sphinx indexes, and using the proper JS query+search code in the tests.

I think it's important enough to block the other search-related PRs until these are in.

**Comment by picnixz** [comment]
> I know there are a lot of these small changes / PRs, so it's probably tricky to develop the context for what's changing. 

Which is the reverse of what I'm  doing on the test suite where the PRs are huge. So don't worry

File: tests/js/fixtures/multiterm/searchindex.js
Comment (picnixz) [inline tests/js/fixtures/multiterm/searchindex.js]: Since it's a fixture, could you make it larger? like more readable? (prettify it). It doesn't hurt for debugging / extending the fixture.

File: tests/js/fixtures/cpp/searchindex.js
Comment (picnixz) [inline tests/js/fixtures/cpp/searchindex.js]: Same comment as above.

File: tests/js/searchtools.js
Comment (picnixz) [inline tests/js/searchtools.js:3]: I never used karma but should I assume that this is the way to do it? namely doing an XMLHttpRequest? isn't it possible just to like.. load the file normally?

File: tests/js/searchtools.js
Comment (picnixz) [inline tests/js/searchtools.js]: You could perhaps create another function called `evalFixture` for that? or is there a more canonical way to do it with karma?

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js:3]: Good question.  So, two considerations: `karma` imports all configured JavaScript files when the tests run, and the `searchindex.js` files affect the state of the application (because it calls `setIndex`).

We want each test case to call `setIndex` individually -- setting state only for that test.

Doing that requires opening and `eval(...)`'ing the fixture.  The tests run in a browser, so they can't open local files -- and so we use the `karma` HTTP server to serve it up for access instead.

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js]: I'll double-check the documentation; perhaps there is a better way (it would seem like importing a script file per-test could be a common use case).

File: tests/js/fixtures/multiterm/searchindex.js
Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: Yep, I was thinking about doing this too.  Especially if/when the format of the index changes, that could be very useful.

To do that, we'd want to edit the Python [`_JavaScriptIndex.dumps` code](https://github.com/sphinx-doc/sphinx/blob/6aaeee206bda210f1b2cd845a924f86b18c0af8e/sphinx/search/__init__.py#L161-L165) to set `indent=4` or something similar.  But we wouldn't want that by default for most projects, because it'll increase the file size of the index.

Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: I think I'd file a feature request for this: a config setting `html_search_minify_index` (or similar), default `True`.  We would set it to `False` in the configuration files in the JS test roots.

File: tests/js/searchtools.js
Comment (picnixz) [reply tests/js/searchtools.js:3]: Oh I see. Good then

File: tests/js/fixtures/multiterm/searchindex.js
Comment (picnixz) [reply tests/js/fixtures/multiterm/searchindex.js]: Wait, the fixtures are always auto-generated? if so, you can just run a post-processing step where you prettify that specific fixture, right?

Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: That could be an easier option, yes!

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js]: I didn't find anything suitable in `karma` itself.  I did find some plugins.. there's [a reasonably popular one](https://github.com/billtrik/karma-fixture) that loads HTML and JSON, but I don't think it is keen on [evaluating JavaScript](https://github.com/billtrik/karma-fixture/issues/19).

I had a look at `jasmine`, too.  We could enable a plugin to provide [mocking of AJAX requests](https://jasmine.github.io/tutorials/mocking_ajax) -- but again given that the code is running in a browser context, then we need to read the fixture data from somewhere.  I'll think about this some more.

File: tests/js/fixtures/multiterm/searchindex.js
Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: The generation occurs within a Python context (running the Sphinx build) so I am considering https://github.com/beautifier/js-beautify which can be installed using `pip`.

Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: (it would be a post-processing step, but even so, we'll want something that can easily run on any developer/CI machine, preferably without installing system-level packages)

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js:3]: Thank you.

File: tests/js/fixtures/multiterm/searchindex.js
Comment (picnixz) [reply tests/js/fixtures/multiterm/searchindex.js]: Ok, for now let's leave it like this

Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: Ah, I missed that comment before developing and pushing the change.

**Comment by jayaddison** [comment]
The latest changes in mainline make a small change to the `searchindex.js` file format.  I'm going to apply a change here too (including merging in a [utility script](https://github.com/jayaddison/sphinx/commit/fbaff49c264640caefd190db3afc10b4cea6e406) that I've written to perform the regeneration and beautification).

File: tests/js/fixtures/multiterm/searchindex.js
Comment (jayaddison) [reply tests/js/fixtures/multiterm/searchindex.js]: > Wait, the fixtures are always auto-generated? if so, you can just run a post-processing step where you prettify that specific fixture, right?

I'm going to remove the post-processing again I think, and there's a specific reason I have in mind: by doing that, I can add a Python test that builds each of these roots and confirms that the `searchindex.js` in the JS test directory is correct.

They should change extremely rarely, and if people want to view prettified diffs of those, they can do relatively easily locally.

File: tests/js/language_data.js
Comment (picnixz) [inline tests/js/language_data.js]:
```suggestion
/* Non-minified version is copied as a separate JS file, if available */
```

Is it a typo?

File: tests/js/searchtools.js
Comment (picnixz) [inline tests/js/searchtools.js:31]: I don't understand why the two tests are now different.

File: utils/generate_js_fixtures.py
Comment (picnixz) [inline utils/generate_js_fixtures.py]: Should we really make it in build or in _build instead?

File: utils/generate_js_fixtures.py
Comment (picnixz) [inline utils/generate_js_fixtures.py]:
```suggestion
SPHINX_ROOT = Path(__file__).resolve().parent.parent
```

(I think it's the correct one, but check it with a print).

File: tests/js/language_data.js
Comment (jayaddison) [reply tests/js/language_data.js]: Does seem to be, yep - thanks!

File: utils/generate_js_fixtures.py
Comment (jayaddison) [reply utils/generate_js_fixtures.py]: Probably better to use `_build` indeed.  Updated (including the `.gitignore` file.  there may be some unrelated stale entries in there to clean up separately).

File: utils/generate_js_fixtures.py
Comment (jayaddison) [reply utils/generate_js_fixtures.py]: That does work, yep - it's less consistent with other scripts in the same directory, so I'm unsure whether to change it.

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js:31]: The closest analogy for this that I can think of is the `srcdir=...` setting for the Python tests.  Each of the JavaScript test cases can (and in the case of the search tests, does) now read from test fixture(s) relevant to that individual test case.

It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

File: utils/generate_js_fixtures.py
Comment (picnixz) [reply utils/generate_js_fixtures.py]: I updated my suggestion since it's better to first resolve and do the parent.parent (in case you have symlinks). It's maybe less consistent but it's clearer. I mean, you need to remember that your parents are in the reverse order whereas with .parent.parent it's clearer to me.

File: tests/js/searchtools.js
Comment (picnixz) [reply tests/js/searchtools.js:31]: Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

Comment (jayaddison) [reply tests/js/searchtools.js:31]: > It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

The relevant `jasmine` guidance to follow if/when doing this appears to be: https://jasmine.github.io/tutorials/sharing_behaviors#sharing-behaviors

Comment (jayaddison) [reply tests/js/searchtools.js:31]: > Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

Good question.  Checking why that happened..

Comment (jayaddison) [reply tests/js/searchtools.js:31]: Ah yes, I forgot about this.  The old test code was only testing a portion of the search query handling - the `performTermsSearch` function.

The duplicate is found during another part of the search code, namely here: https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L323-L338

(note when navigating further down that code file that [`performTermsSearch` is called later](https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L367-L368))

When we update the code here to use the top-level `_parseQuery` function -- that encompasses all of the query-handling code -- the duplicate appears in the results because we check both `term` and `title` matches.

Comment (picnixz) [reply tests/js/searchtools.js:31]: Could you make the old test compatible with your fixtures as well? it'd be good to test specific components.

Comment (jayaddison) [reply tests/js/searchtools.js:31]: To check I understand you: partially revert the changes to the `should be able to search for multiple terms`, so that it continues to load a search index from fixtures, and only tests the `performTermsSearch` function, as-before.

Comment (picnixz) [reply tests/js/searchtools.js:31]: With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:

- keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
- you create a new test for testing what you just added (performing the full E2E, with the duplicates).

Comment (jayaddison) [reply tests/js/searchtools.js:31]: Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

Comment (jayaddison) [reply tests/js/searchtools.js:31]: > I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

Yep, let's do that first, if that's OK?  I've opened #12229 for it.

File: tests/js/searchtools.js
Comment (jayaddison) [inline tests/js/searchtools.js:13]: This test case has begun failing, and the reason is that the token `C++` (as found in a source document) is typically reduced down to the single-character term `c` in the search index.

I think it's best to resolve this by using `Search.parseQuery` to form a query token that will match `c`, instead of trying to get a literal term `c++` into the index (which is possible, but perhaps not best practice).

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js:31]: > > With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:
> >
> >  * keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
> >
> >  * you create a new test for testing what you just added (performing the full E2E, with the duplicates).
> 
> Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

Ok, agreed - I think it is OK to have both types of test in place, and I've pushed corresponding changes.

File: tests/test_search.py
Comment (jayaddison) [inline tests/test_search.py]: Some caution about these lines: I _thought_ that using `pathlib.Path` files' `read_bytes()` and hashing the results of that should produce identical results on different platforms.

A [test failure](https://github.com/sphinx-doc/sphinx/actions/runs/8563318800/job/23468161852#step:6:2342) on Windows where differences appear with very similar code (from an otherwise-unrelated PR) is making me want to double-check that.

Comment (jayaddison) [reply tests/test_search.py]: Ok, in the case of `searchindex.js` this appears not to be a problem - I believe due to the absence of newlines in the file content (meaning that the byte-for-byte content of the file is unchanged when written on different operating system platforms).

**Comment by jayaddison** [comment]
Thanks @wlach!  I think I should re-test this manually after the latest merge; any additional code review feedback would be useful.  I do remember that there was some JavaScript variable assignment logic that I felt was a bit non-obvious (see the within-line comments attempting to explain it).

**Comment by jayaddison** [comment]
> Thanks @wlach! I think I should re-test this manually after the latest merge; any additional code review feedback would be useful. I do remember that there was some JavaScript variable assignment logic that I felt was a bit non-obvious (see the within-line comments attempting to explain it).

Ok, yep - manually testing from 9f5df2cccaf41e81b58fee9a47bf98edfbe798cf seems good here; I took the opportunity to merge/re-test #12047 after doing that, and it tests correctly too (the `Welcome` search test case for `sphinx.git` de-duplicates two top search results into one with the fix in place; the search results for `configuration` continue to be a bit spammy, but that doesn't appear to be due to title-based duplication).

File: CHANGES.rst
Comment (wlach) [inline CHANGES.rst]: This needs to be moved up to the current release. Don't forget to credit yourself!

File: tests/js/language_data.js
Comment (wlach) [inline tests/js/language_data.js:6]: The comments in this file seem a little misleading? This is really just test scaffolding AFAICT

File: tests/js/roots/multiterm/index.rst
Comment (wlach) [inline tests/js/roots/multiterm/index.rst]: I would personally use this opportunity to add a bit of context about what this file tests.

File: tests/js/roots/partial/index.rst
Comment (wlach) [inline tests/js/roots/partial/index.rst]: Same, I would personally use this opportunity to add a bit of context about what this file tests.

File: tests/js/searchtools.js
Comment (wlach) [inline tests/js/searchtools.js:6]: I'm guessing you used `XMLHttpRequest()` because you want to make a synchronous? Should probably add a comment about that.

Apparently using [XMLHttpRequest this way is discouraged](https://stackoverflow.com/a/72561702/295132) and might be deprecated at some point, but they've been saying that since 2014. 😉 It's probably fine.

File: tests/js/searchtools.js
Comment (wlach) [inline tests/js/searchtools.js]: I would personally use underscores in these cases but it's not that big a deal.

```suggestion
      [_, searchterms, excluded, _] = Search._parseQuery('sphinx');
```

File: tests/test_search.py
Comment (wlach) [inline tests/test_search.py]: It's a bit unclear to me why you're taking a hash of the file rather than just comparing the contents directly? Feels like that would be just as fast and probably simpler given the sizes of our files.

**Review by wlach — COMMENTED** [review]
Nice work! I left a few comments which are mostly just nits.

File: tests/js/language_data.js
Comment (jayaddison) [reply tests/js/language_data.js:6]: Mm, yep - this is an interesting case - it's an evaluated copy of a [template file](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/themes/basic/static/language_data.js_t) from the basic theme.  That's me attempting to make the JavaScript test environment realistic, but it is duplicative (and a bit confusing).

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js:6]: Ok, that's good to know :)  Yep; we don't want the test to proceed until the fixture has definitely loaded -- I'm inclined to use a straightforward (even if seemingly dated) approach here.

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js]: JavaScript's unpacking behaviour here seems somewhat inconsistent to me; the value on the left is a single item from the value on the right-hand-side, and yet the last value (the second `_` in your suggestion) is assigned all of the remaining items.  I was hoping for a way to retain that information for future readers/copiers in case they attempted to make adjustments that wouldn't work as they expect.

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py]: Initially I felt some concern about this code perhaps being reused for larger files -- to the extent that entire-file comparison might not be ideal -- however, it is true that the search indexes are small, and so the simplicity benefit of comparing the full content does seem worthwhile.

**Comment by jayaddison** [comment]
Thank you @wlach for the code review!  I've left a couple of not-entirely-resolved items open, and let me know what you think of the adjustments.

File: tests/js/searchtools.js
Comment (wlach) [reply tests/js/searchtools.js]: Ah ok, you learn something new every day. ;) I might use an underscore just for the first element then. But as I said above, not a big deal.

**Review by wlach — APPROVED** [review]
Looks good from my perspective! Obviously you'll need a committer to approve/land.

**Comment by wlach** [comment]
@picnixz Could you take a look at this? I think it makes sense to merge this before proceeding with any more work on the search system. I reviewed it a couple weeks ago and all looks good to me (not to say I'm infallible but that's gotta count for something)

File: utils/generate_js_fixtures.py
Comment (picnixz) [inline utils/generate_js_fixtures.py]: Can you use full options here ? It'd be a bit clearer in general (I think we use full options in CI/CD so that's why).

File: tests/test_search.py
Comment (picnixz) [inline tests/test_search.py]:
```suggestion
JAVASCRIPT_TEST_ROOTS = list((TESTS_ROOT / 'js' / 'roots').iterdir())
@pytest.mark.parametrize('directory', JAVASCRIPT_TEST_ROOTS)
```

For parametrized tests, I prefer having a non-iterator to ensure that we have everything *before* launching the test.

File: tests/js/searchtools.js
Comment (picnixz) [reply tests/js/searchtools.js]: > and yet the last value (the second _ in your suggestion) is assigned all of the remaining items

Technically, no. It would be assigned the remaining items if you were using `...rest`. You can check:

```javascript
[_, two, _] = [1,2,3,4,5]
console.log(`${_}`) // 3

[_, two, ..._] = [1,2,3,4,5]
console.log(`${_}`) // 3,4,5
```

It's the same as in Python:

```python
(one, two, *threefourfive) = (1, 2, 3, 4, 5)
```

File: tests/js/fixtures/partial/searchindex.js
Comment (picnixz) [inline tests/js/fixtures/partial/searchindex.js:1]: I don't remember whether I told/asked you or not, but do we need a blank line at the end of this file (and the others)?

image

**Comment by picnixz** [comment]
> Could you take a look at this? I think it makes sense to merge this before proceeding with any more work on the search system

Lucky you that I've submitted my thesis and have more free time!

File: tests/js/searchtools.js
Comment (jayaddison) [reply tests/js/searchtools.js]: Ok, yep - I misunderstood JavaScript's unpacking behaviour.  I'll adjust this to use meaningful-and-underscore-prefixed variable names, and with unpacking for the remainder items.

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py]: Although it may be non-obvious, this should evaluate and construct a non-iterator value for `JAVASCRIPT_TEST_ROOTS` -- the `iterdir()` call is wrapped within a `list(...)` call that should evaluate it.  (or maybe I misunderstood the suggestion here?)

File: tests/js/fixtures/partial/searchindex.js
Comment (jayaddison) [reply tests/js/fixtures/partial/searchindex.js:1]: As output by Sphinx itself, the `searchindex.js` file does not include an end-of-file newline, and since the fixtures are generated by Sphinx itself, I'd prefer that they don't contain those either.

At one point I did dabble with adding auto-reformatting of the fixtures -- arguably it could help when reviewing diffs that affect the index structure -- but on balance I preferred the simplicity of not adding any post-processing steps (and dependencies) for the JS file (adding some emphasis there: the index file is JavaScript code, not JSON data).

File: tests/test_search.py
Comment (picnixz) [reply tests/test_search.py]: Ma suggestion is to have a list in memory when calling `parametrize` and not a generator constructed on the fly. Here,

```
(directory for directory in (TESTS_ROOT / 'js' / 'roots').iterdir())
``` 

is a generator (not a tuple!) (so it is still lazily evaluated) and actually equivalent to `(TESTS_ROOT / 'js' / 'roots').iterdir()`. In general, pytest recommends to have eager values and not iterated values (for instance, if you use my suggestion, it's easy to overwrite the value of JAVASCRIPT_TEST_ROOTS).

**Review by picnixz — APPROVED** [review]
For me it's fine now (modulo this small suggestion but we can just decide later).

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py]: Argh, my mistake.  Somehow I read your code review suggestion (that _does_ include a `list(...)` call) as the current state of the test code.  However, indeed, as-is this branch does use an iterator.  Fixing this momentarily..

**Comment by picnixz** [comment]
Once you're done, just ping me and I'll merge it !

**Comment by jayaddison** [comment]
Thanks @picnixz - all done with those changes, so I think this is ready for merge.

**Comment by picnixz** [comment]
Thank you!

**Pull request merged**
